orchestration/flows/scicat/ingestor_controller.py

```diff
@@ -0,0 +1,328 @@
+
+        logger.info(f"Logging in to SciCat at {scicat_base_url} as {scicat_user}.")
+
+        # Ensure that all required credentials are provided.
+        if not (scicat_base_url and scicat_user and scicat_password):
+            raise ValueError(
+                "Missing required SciCat credentials. Provide scicat_base_url, scicat_user, "
+                "and scicat_password as parameters or set them in the environment variables: "
+                "SCICAT_API_URL, SCICAT_INGEST_USER, SCICAT_INGEST_PASSWORD."
+            )
+
+        # Try to log in using the pyscicat client first.
+        # This method seems deprecated, but leaving it here for backwards compatability
+        # https://github.com/SciCatProject/pyscicat/issues/61
+        try:
+            self.scicat_client = from_credentials(
+                base_url=scicat_base_url,
+                username=scicat_user,
+                password=scicat_password
+            )
+            logger.info("Logged in to SciCat.")
+            return self.scicat_client
+        except Exception as e:
+            logger.warning(f"Failed to log in to SciCat: {e}, trying alternative method.")
+
+        # This method works for scicatlive 3.2.5
+        try:
+            url = urljoin(scicat_base_url, "auth/login")
+            logger.info(url)
+            response = requests.post(
+                url=url,
+                json={"username": scicat_user, "password": scicat_password},
+                stream=False,
+                verify=True,
+            )
+            logger.info(f"Login response: {response.json()}")
+            self.scicat_client = ScicatClient(scicat_base_url, response.json()["access_token"])
+            logger.info("Logged in to SciCat.")
+            # logger.info(f"SciCat token: {response.json()['access_token']}")
+            return self.scicat_client
+
+        except requests.exceptions.RequestException as e:
+            logger.error(f"Failed to log in to SciCat: {e}")
+            raise e
+        except Exception as e:
+            logger.error(f"Failed to log in to SciCat: {e}")
+            raise e
+
+    @abstractmethod
+    def ingest_new_raw_dataset(
+        self,
+        file_path: str = "",
+    ) -> str:
+        """Ingest data from the beamline.
+
+        :param file_path: Path to the file to ingest.
+        :return: SciCat ID of the dataset.
+        """
+        pass
+
+    @abstractmethod
+    def ingest_new_derived_dataset(
+        self,
+        file_path: str = "",
+        raw_dataset_id: str = "",
+    ) -> str:
+        """Ingest data from the beamline.
+
+        :param file_path: Path to the file to ingest.
+        :return: SciCat ID of the dataset.
+        """
+        pass
+
+    def add_new_dataset_location(
+        self,
+        dataset_id: Optional[str] = None,
+        proposal_id: Optional[str] = None,
+        file_name: Optional[str] = None,
```

Comment:
There are three optional parameters, but you need either dataset_id or proposal_id+file_name. I think it would be cleaner if this method required dataset_id, and the find dataset were more of a utility funciton for those use cases where it might not be known.